### PR TITLE
(polyline): add some new features

### PR DESCRIPTION
**Pull request by caroph**
Polylines: Add new features on highlight and alter events 
Refac: make code easier to maintain

### CODE REVIEW

File: src/models/apis/google/google-markers.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        markerClusterer.clearMarkers();
+    }
+
+    public countMarkersOnCluster(markerClusterer: any): number {
+        return markerClusterer.getMarkers().length;
+    }
+
+    private moveTransitionMarker(position: any, marker: any) {
+        const numDeltas = 5;
+        const referencia = {
```
Comment (lidia-freitas) [inline src/models/apis/google/google-markers.ts]: pt-br?

File: src/models/apis/google/google-markers.ts [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            i: 0,
+            deltaLat: (position.lat - marker.getPosition().lat()) / numDeltas,
+            deltaLng: (position.lng - marker.getPosition().lng()) / numDeltas
+        };
+
+        this.moveMarker(marker, referencia, numDeltas);
+    }
+
+    private moveMarker(marker: any, referencia: any, numDeltas: number) {
+        referencia.position[0] += referencia.deltaLat;
```
Comment (lidia-freitas) [inline src/models/apis/google/google-markers.ts]: pt-br?

File: src/models/apis/google/google-polylines.ts [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+            polyline.navigationHandlerClick();
+        }
+    }
+
+    private onKeyUp(event: any) {
+        const self = this;
+
+        if (self.selectedPath) {
+            switch (event.which ? event.which : event.keyCode) {
+                // seta para cima ou seta para direita ou W ou D
```
Comment (lidia-freitas) [inline src/models/apis/google/google-polylines.ts]: pt-br

File: src/models/apis/google/google-polylines.ts [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+        const deltaSum = (deltaX * deltaX) + (deltaY * deltaY);
+
+        incIntersect += (pt.lat() - pt1.lat()) * deltaY;
+        if (deltaSum > 0) {
+            incIntersect /= deltaSum;
+        } else {
+            incIntersect = -1;
+        }
+
+        // A interseção ocorre fora do segmento de reta, 'antes' do pt1.
```
Comment (lidia-freitas) [inline src/models/apis/google/google-polylines.ts]: pt-br

File: src/models/apis/leaflet/leaflet-geojson.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                return new this.leaflet.Marker(parsedCoordinates, objectOptions);
+            case 'Polygon':
+                geometry.coordinates
+                    .forEach((polygon) => parsedCoordinates.push(polygon.map((elem) => elem.reverse())));
+                return new this.leaflet.Polygon(parsedCoordinates, objectOptions);
+            case 'LineString':
+                parsedCoordinates = geometry.coordinates.map((elem) => elem.reverse());
+                return new this.leaflet.Polyline(parsedCoordinates, objectOptions);
+            default:
+                throw new Error('Forma de objeto desconhecida.');
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-geojson.ts]: pt-br

File: src/models/apis/leaflet/leaflet-markers.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        markerClusterer.clearLayers();
+    }
+
+    public countMarkersOnCluster(markerClusterer: any): number {
+        return markerClusterer.getLayers().length;
+    }
+
+    private moveTransitionMarker(position: any, marker: any) {
+        const numDeltas = 5;
+        const referencia = {
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-markers.ts]: pt-br

File: src/models/apis/leaflet/leaflet-markers.ts [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            deltaLat: (position[0] - marker.getLatLng().lat) / numDeltas,
+            deltaLng: (position[1] - marker.getLatLng().lng) / numDeltas
+        };
+
+        this.moveMarker(marker, referencia, numDeltas);
+    }
+
+    private moveMarker(marker: any, referencia: any, numDeltas: number) {
+        referencia.position[0] += referencia.deltaLat;
+        referencia.position[1] += referencia.deltaLng;
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-markers.ts]: pt-br

File: src/models/apis/leaflet/leaflet-polylines.ts [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+    private onKeyUp(event: any) {
+        const self = this;
+
+        if (self.selectedPath) {
+            if (event.ctrlKey) {
+                console.warn('Deprecated: CTRL is not needed in navigation anymore');
+            }
+
+            switch (event.which ? event.which : event.keyCode) {
+                // seta para cima ou seta para direita ou W ou D
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-polylines.ts]: pt-br

File: src/models/apis/leaflet/leaflet-polylines.ts [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+                case 38:
+                case 39:
+                    self.moveFowards(event.shiftKey);
+                    break; // seta para esquerda ou seta para baixo ou S ou A
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-polylines.ts]: pt-br

File: src/models/apis/leaflet/leaflet-polylines.ts [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+        const deltaSum = (deltaX * deltaX) + (deltaY * deltaY);
+
+        incIntersect += (pt.lat - pt1.lat) * deltaY;
+        if (deltaSum > 0) {
+            incIntersect /= deltaSum;
+        } else {
+            incIntersect = -1;
+        }
+
+        // A interseção ocorre fora do segmento de reta, "antes" do pt1.
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-polylines.ts]: pt-br

File: src/models/apis/leaflet/leaflet-polylines.ts [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+        if (incIntersect < 0) {
+            return self.kmTo(pt, pt1);
+        } else if (incIntersect > 1) {
+            return self.kmTo(pt, pt2);
+        }
+
+        // Cálculo do ponto de interseção.
```
Comment (lidia-freitas) [inline src/models/apis/leaflet/leaflet-polylines.ts]: pt-br

File: src/models/apis/google/google-markers.ts (hunk 1 above)
Comment (caroph) [reply src/models/apis/google/google-markers.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/google/google-markers.ts (hunk 2 above)
Comment (caroph) [reply src/models/apis/google/google-markers.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/google/google-polylines.ts (hunk 3 above)
Comment (caroph) [reply src/models/apis/google/google-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/google/google-polylines.ts (hunk 4 above)
Comment (caroph) [reply src/models/apis/google/google-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-geojson.ts (hunk 5 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-geojson.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-markers.ts (hunk 6 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-markers.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-markers.ts (hunk 7 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-markers.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-polylines.ts (hunk 8 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-polylines.ts (hunk 9 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-polylines.ts (hunk 10 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

File: src/models/apis/leaflet/leaflet-polylines.ts (hunk 11 above)
Comment (caroph) [reply src/models/apis/leaflet/leaflet-polylines.ts]: resolvido no commit e8277315b503e7b2710ce1d332c9a43d8c072481

**Pull request merged**